astropy/utils/shapes.py

```diff
@@ -298,49 +284,32 @@ def self_iter():
         for name in np_core.fromnumeric.__all__
         if name not in ["alen", "sort", "partition"]
     }
-    # Add np.copy, which we may as well let defer to our method.
     _METHOD_FUNCTIONS[np.copy] = "copy"
 
-    # Could be made to work with a bit of effort:
-    # np.where, np.compress, np.extract,
-    # np.diag_indices_from, np.triu_indices_from, np.tril_indices_from
-    # np.tile, np.repeat (need .repeat method)
-    # TODO: create a proper implementation.
-    # Furthermore, some arithmetic functions such as np.mean, np.median,
-    # could work for Time, and many more for TimeDelta, so those should
-    # override __array_function__.
     def __array_function__(self, function, types, args, kwargs):
         """Wrap numpy functions that make sense."""
-        if function in self._APPLICABLE_FUNCTIONS:
-            if function is np.broadcast_to:
-                # Ensure that any ndarray subclasses used are
-                # properly propagated.
-                kwargs.setdefault("subok", True)
-            elif (
-                function in {np.atleast_1d, np.atleast_2d, np.atleast_3d}
-                and len(args) > 1
-            ):
-                return tuple(function(arg, **kwargs) for arg in args)
-
-            if self is not args[0]:
-                return NotImplemented
-
-            return self._apply(function, *args[1:], **kwargs)
-
-        # For functions that defer to methods, use the corresponding
-        # method/attribute if we have it.  Otherwise, fall through.
-        if self is args[0] and function in self._METHOD_FUNCTIONS:
-            method = getattr(self, self._METHOD_FUNCTIONS[function], None)
-            if method is not None:
-                if callable(method):
-                    return method(*args[1:], **kwargs)
-                else:
-                    # For np.shape, etc., just return the attribute.
-                    return method
-
-        # Fall-back, just pass the arguments on since perhaps the function
-        # works already (see above).
-        return function.__wrapped__(*args, **kwargs)
+        if not all(issubclass(t, ShapedLikeNDArray) for t in types):
+            return NotImplemented
+
+        namespace = self.__array_namespace__()
+        if function in namespace:
+            return namespace[function](*args, **kwargs)
+        return NotImplemented
```

Comment:
I don't think it's this simple. I think this change erases a lot of required logic above.